scheduler/pkg/server/pipeline_status.go

```diff
@@ -104,128 +172,176 @@ func (s *SchedulerServer) sendCurrentPipelineStatuses(
-	for _, pipelineEventMessage := range pipelineEventMessages {
-		runningPipelines = append(runningPipelines, pipelineEventMessage.PipelineName)
-	}
-	return runningPipelines
-}
-
-func (s *SchedulerServer) createPipelineDeletionMessage(pip *pipeline.Pipeline, keepTopics bool) (*pb.PipelineStatusResponse, error) {
+func (s *SchedulerServer) createPipelineDeletionMessage(pv *pipeline.PipelineVersion) *pb.PipelineStatusResponse {
 	return &pb.PipelineStatusResponse{
-		PipelineName: pip.Name,
-		Versions:     []*pb.PipelineWithState{},
-	}, nil
+		Operation:    pb.PipelineStatusResponse_PipelineDelete,
+		PipelineName: pv.Name,
+		Versions: []*pb.PipelineWithState{
+			pipeline.CreatePipelineWithState(pv),
+		},
+	}
 }
 
-func (s *SchedulerServer) createPipelineCreationMessage(pip *pipeline.Pipeline) (*pb.PipelineStatusResponse, error) {
-	pipelineVersion := pip.GetLatestPipelineVersion()
-	pipelineWithState := pipeline.CreatePipelineWithState(pipelineVersion)
+func (s *SchedulerServer) createPipelineCreationMessage(pv *pipeline.PipelineVersion) *pb.PipelineStatusResponse {
 	return &pb.PipelineStatusResponse{
-		PipelineName: pip.Name,
-		Versions:     []*pb.PipelineWithState{pipelineWithState},
-	}, nil
+		Operation:    pb.PipelineStatusResponse_PipelineCreate,
+		PipelineName: pv.Name,
+		Versions: []*pb.PipelineWithState{
+			pipeline.CreatePipelineWithState(pv),
+		},
+	}
 }
 
 func (s *SchedulerServer) pipelineGwRebalance() {
 	s.pipelineEventStream.mu.Lock()
 	defer s.pipelineEventStream.mu.Unlock()
 
-	runningPipelines := s.GetAllRunningPipelines()
-	s.logger.Debugf("Rebalancing pipeline gateways for running pipelines: %v", runningPipelines)
+	// get only the pipeline gateway streams
+	streams := []*PipelineSubscription{}
+	for _, subscription := range s.pipelineEventStream.streams {
+		if subscription.isPipelineGateway {
+			streams = append(streams, subscription)
+		}
+	}
+
+	evts := s.pipelineHandler.GetAllPipelineGwRunningPipelineVersions()
+	for _, event := range evts {
+		pv, err := s.pipelineHandler.GetPipelineVersion(event.PipelineName, event.PipelineVersion, event.UID)
+		if err != nil {
+			s.logger.WithError(err).Errorf("Failed to get pipeline version for %s:%d (%s)", event.PipelineName, event.PipelineVersion, event.UID)
+			continue
+		}
 
-	for _, pipelineName := range runningPipelines {
-		pip, _ := s.pipelineHandler.GetPipeline(pipelineName)
-		consumerBucketId := s.getPipelineGatewayBucketId(pipelineName)
-		servers := s.pipelineGWLoadBalancer.GetServersForKey(consumerBucketId)
-		s.logger.Debugf("Servers for pipeline %s: %v", pipelineName, servers)
-		s.logger.Debug("Consumer bucket ID: ", consumerBucketId)
+		s.logger.Debugf(
+			"Rebalancing pipeline %s:%d with pipeline gateway state %s",
+			event.PipelineName, event.PipelineVersion, pv.State.PipelineGwStatus.String(),
+		)
 
-		// need to update envoy clusters
+		// no pipeline gateway available yet, publish event for envoy
```

Comment:
How do we know at this point that there are no pipeline gateways available at this point? (I was just slightly confused reading the comment)